### PR TITLE
Add D4 pool deployment scripts

### PR DESCRIPTION
**Pull request by alphastorm**
Uses latest `SwapUtils`, `Swap`, `SwapFlashLoan`, `SwapDeployer`, `LPToken` contracts.

Please double check `alUSD` / `FEI` / `FRAX` / `LUSD` have the correct addresses and decimals.

### CODE REVIEW

File: deploy/999_impersonate_owners.ts [hunk 1]
```diff
@@ -30,6 +31,11 @@ const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
     ALETH: ["0x1d2c4cd9bee9dfe088430b95d274e765151c32db"],
     WETH: ["0xceff51756c56ceffca006cd410b03ffc46dd3a58"],
     SETH: ["0xc5424b857f758e906013f3555dad202e4bdb4567"],
+    // D4
+    ALUSD: ["0x43b4fdfd4ff969587185cdb6f0bd875c5fc83f8c"],
```
Comment (hammeiam) [inline deploy/999_impersonate_owners.ts:35]: this appears to be the alUSD3crv address https://etherscan.io/address/0x43b4fdfd4ff969587185cdb6f0bd875c5fc83f8c. The correct address is `0xbc6da0fe9ad5f3b0d58160288917aa56653660e9`

File: deploy/999_impersonate_owners.ts [hunk 2]
```diff
@@ -30,6 +31,11 @@ const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
...
+    FEI: ["0x94b0a3d511b6ecdb17ebf877278ab030acb0a878"],
```
Comment (hammeiam) [inline deploy/999_impersonate_owners.ts:36]: this appears to be the FEI/ETH uniswap pool https://etherscan.io/address/0x94b0a3d511b6ecdb17ebf877278ab030acb0a878, correct address is `0x956F47F50A910163D8BF957Cf5846D573E7f87CA`

File: deployments/mainnet/FRAX.json [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+          "indexed": false,
+          "name": "value",
+          "type": "uint256"
+        }
+      ],
+      "name": "Transfer",
+      "type": "event"
+    }
+  ],
+  "address": "0x853d955aCEf822Db058eb8505911ED77F175b99e"
```
Comment (hammeiam) [inline deployments/mainnet/FRAX.json:224]: lgtm

File: deployments/mainnet/FEI.json [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+          "indexed": false,
+          "name": "value",
+          "type": "uint256"
+        }
+      ],
+      "name": "Transfer",
+      "type": "event"
+    }
+  ],
+  "address": "0x956F47F50A910163D8BF957Cf5846D573E7f87CA"
```
Comment (hammeiam) [inline deployments/mainnet/FEI.json:224]: lgtm

File: deployments/mainnet/LUSD.json [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+          "indexed": false,
+          "name": "value",
+          "type": "uint256"
+        }
+      ],
+      "name": "Transfer",
+      "type": "event"
+    }
+  ],
+  "address": "0x5f98805A4E8be255a32880FDeC7F6728C6568bA0"
```
Comment (hammeiam) [inline deployments/mainnet/LUSD.json:224]: lgtm

File: deployments/mainnet/ALUSD.json [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+          "indexed": false,
+          "name": "value",
+          "type": "uint256"
+        }
+      ],
+      "name": "Transfer",
+      "type": "event"
+    }
+  ],
+  "address": "0xBC6DA0FE9aD5f3b0d58160288917AA56653660E9"
```
Comment (hammeiam) [inline deployments/mainnet/ALUSD.json:224]: lgtm

File: deploy/150_deploy_D4Pool.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    log(`reusing "SaddleD4Pool" at ${SaddleD4Pool.address}`)
+  } else {
+    // Constructor arguments
+    const TOKEN_ADDRESSES = [
+      (await get("ALUSD")).address,
+      (await get("FEI")).address,
+      (await get("FRAX")).address,
+      (await get("LUSD")).address,
+    ]
+    const TOKEN_DECIMALS = [18, 18, 18, 18]
```
Comment (hammeiam) [inline deploy/150_deploy_D4Pool.ts:21]: lgtm

File: deploy/150_deploy_D4Pool.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+      ).address,
+    )
+
+    const newPoolEvent = receipt?.events?.find(
+      (e: any) => e["event"] == "NewSwapPool",
+    )
+    const D4SwapAddress = newPoolEvent["args"]["swapAddress"]
+    log(`deployed D4 pool (targeting "SwapFlashLoan") at ${D4SwapAddress}`)
+    await save("SaddleD4Pool", {
+      abi: (await get("SwapFlashLoan")).abi,
```
Comment (hammeiam) [inline deploy/150_deploy_D4Pool.ts:53]: double checking that this is the most up to date with gas savings etc?

**Review by hammeiam — APPROVED** [review]
addresses and decimals look good!

**Review by penandlim — APPROVED** [review]
LGTM!

File: deploy/150_deploy_D4Pool.ts (hunk 8 above)
Comment (penandlim) [reply deploy/150_deploy_D4Pool.ts:53]: Yep looks good to me.

**Review by sandrasong — APPROVED** [review]

**Pull request merged**